### PR TITLE
UTY-314: Docs Linting

### PR DESCRIPTION
**Pull request by jamiebrynes7**
#### Description
Added a docs linter to the Unity GDK. It currently will check links to verify that they point to valid locations. It will check:
* Any in-file links - `#documentation`
* Any other-markdown file links - `../README.md#documentation`
* Any non-markdown file links - `../my-image.png`
* Any remote URLs - `https://www.google.com`

It should be ran in the root of the project and will check all `*.md` files in the entire project.

Note that for the remote URL checking, it will issue a `GET` request to each URL. Any `404` return codes are treated as a fail, `200` is treated as a success, any other status code prints a warning but **does not fail**. 

~**Left To-Do:**~

1. ~Add doc-strings to the classses/methods~
2. ~Write something about it in the docs~
3. ~Add some tests~

Note that I had to fiddle around the ci scripts slightly as `DocsLinter` relies on `ImpNuget` currently but they both exist in the `Tools.sln` project.

Some example output: 
```
+ bin/DocsLinter/DocsLinter.exe
Checking links for: C:\Workspace\Develop\UnityGDK\CONTRIBUTING.md
Checking links for: C:\Workspace\Develop\UnityGDK\LICENSE.md
Checking links for: C:\Workspace\Develop\UnityGDK\README.md
Error in C:\Workspace\Develop\UnityGDK\README.md. The link: https://docs.improbable.io/reference/13.0/shared/concepts/spatialo is invalid
Error in C:\Workspace\Develop\UnityGDK\README.md. The link: docs/README.md#documentaion is invalid
Error in C:\Workspace\Develop\UnityGDK\README.md. The link: docs/setup-and-instaling.md is invalid
Checking links for: C:\Workspace\Develop\UnityGDK\.github\ISSUE_TEMPLATE.md
Checking links for: C:\Workspace\Develop\UnityGDK\.github\pull_request_template.md
Checking links for: C:\Workspace\Develop\UnityGDK\code_generator\packages\NUnit.3.10.1\CHANGES.md
Checking links for: C:\Workspace\Develop\UnityGDK\docs\README.md
Checking links for: C:\Workspace\Develop\UnityGDK\docs\setup-and-installing.md
Warning in C:\Workspace\Develop\UnityGDK\docs\setup-and-installing.md. The link https://console.improbable.io returned a status code of: 401
Checking links for: C:\Workspace\Develop\UnityGDK\docs\content\authority.md
Checking links for: C:\Workspace\Develop\UnityGDK\docs\content\code-generator.md
Checking links for: C:\Workspace\Develop\UnityGDK\docs\content\commands.md
Checking links for: C:\Workspace\Develop\UnityGDK\docs\content\component-data.md
Checking links for: C:\Workspace\Develop\UnityGDK\docs\content\create-entity.md
Checking links for: C:\Workspace\Develop\UnityGDK\docs\content\custom-replication-system.md
Checking links for: C:\Workspace\Develop\UnityGDK\docs\content\entity-checkout-process.md
Checking links for: C:\Workspace\Develop\UnityGDK\docs\content\events.md
Checking links for: C:\Workspace\Develop\UnityGDK\docs\content\reactive-components.md
Checking links for: C:\Workspace\Develop\UnityGDK\docs\contributions\unity-gdk-coding-standards.md
```

#### Tests
Ran `ci/docs-lint.sh` from a clean state and found an error! Fixed the error and it passed. Need to add in full tests using NUnit.
#### Documentation
Added to Wiki
#### Primary reviewers
If your change will take a long time to review, you can name at most two primary reviewers who are ultimately responsible for reviewing this request. @ mention them.
@jared-improbable @samiwh

### CODE REVIEW

File: tools/DocsLinter/Program.cs [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+{
+    internal class Program
+    {
+        private static void Main(string[] args)
+        {
+            if (args.Contains("--help") || args.Contains("/?") || args.Contains("/help"))
+            {
+                Console.WriteLine(
+                    "A cross-platform helper that lints markdown documents in the current working directory.");
+                Console.WriteLine("This linter currently supports checking local links and images.");
```
[the remark this answers is not in the document]
Comment (jamiebrynes7) [reply tools/DocsLinter/Program.cs]: Done

File: ci/test.sh [hunk 2]
```diff
@@ -15,12 +15,20 @@ markStartOfBlock "Setup variables"
 PROJECT_DIR="$(pwd)"
 mkdir -p "${PROJECT_DIR}/logs/"
 
+TOOLS_TEST_RESULTS_FILES="${PROJECT_DIR}/logs/code-generator-test-results.xml"
```
Comment (jared-improbable) [inline ci/test.sh]: Filename dupes the one below

File: ci/test.sh [hunk 3]
```diff
@@ -63,6 +71,10 @@ PLAYMODE_TEST_RESULT=$?
 
 markEndOfBlock "Playmode Testing"
 
+if [ $TOOLS_TEST_RESULT -ne 0 ]; then
+    >&2 echo "Code Generator Tests failed. Please check the file ${TOOLS_TEST_RESULTS_FILES} for more information."
```
Comment (jared-improbable) [inline ci/test.sh]: Message dupes the one below

File: docs/contributions/unity-gdk-coding-standards.md [hunk 4]
```diff
@@ -9,7 +9,7 @@
 
 _[&copy; 2018 xkcd](https://xkcd.com/license.html)_
 
-**Contributions**: We are not currently taking public contributions - see our [contributions](../../README.md#contributions) policy. However, we are accepting issues and we do want your [feedback](../../README.md#feedback).
+**Contributions**: We are not currently taking public contributions - see our [contributions](../../README.md#contributions) policy. However, we are accepting issues and we do want your [feedback](../../README.md#give-us-feedback).
```
Comment (jared-improbable) [inline docs/contributions/unity-gdk-coding-standards.md]: Drive-by?

File: tools/DocsLinter/DocsLinter.csproj [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    <Reference Include="Markdig, Version=0.15.0.0, Culture=neutral, processorArchitecture=MSIL">
+      <HintPath Condition="Exists('packages')">packages\Markdig.0.15.0\lib\net40\Markdig.dll</HintPath>
+      <HintPath Condition="Exists('..\packages')">..\packages\Markdig.0.15.0\lib\net40\Markdig.dll</HintPath>
+    </Reference>
+    <Reference Include="nunit.framework, Version=3.4.0.0, Culture=neutral, PublicKeyToken=2638cd05610744eb, processorArchitecture=MSIL">
+      <SpecificVersion>False</SpecificVersion>
+      <HintPath Condition="Exists('packages')">packages\NUnit.3.10.1\lib\net45\nunit.framework.dll</HintPath>
+      <HintPath Condition="Exists('..\packages')">..\packages\NUnit.3.10.1\lib\net45\nunit.framework.dll</HintPath>
+    </Reference>
+    <Reference Include="System" />
```
Comment (jared-improbable) [inline tools/DocsLinter/DocsLinter.csproj]: Cleanup unused references

File: tools/DocsLinter/Program.cs [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+        /// </summary>
+        /// <param name="markdownFilePath">The fully qualified path of the Markdown file to check</param>
+        /// <param name="markdownFileContents">An object representing the Markdown file to check.</param>
+        /// <param name="markdownFiles">The corpus of Markdown files undergoing linting. Maps filename to Markdown file object representation.</param>
+        /// <returns>A bool indicating the success of the check.</returns>
+        internal static bool CheckMarkdownFile(string markdownFilePath, SimplifiedMarkdownDoc markdownFileContents,
+            Dictionary<string, SimplifiedMarkdownDoc> markdownFiles)
+        {
+            var allLinksValid = true;
+            foreach (var link in markdownFileContents.Links)
```
Comment (jared-improbable) [inline tools/DocsLinter/Program.cs]: `foreach(... .Links.OfType<LocalLink>()) { }`

File: tools/DocsLinter/Program.cs [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+                    {
+                        LogInvalidLink(markdownFilePath, remoteLink);
+                        return false;
+                    }
+
+                    LogLinkWarning(markdownFilePath, remoteLink, $"returned a status code of: {(int)statusCode}");
+                    return true;
+                }
+
+                Console.Error.WriteLine(ex.ToString());
```
Comment (jared-improbable) [inline tools/DocsLinter/Program.cs]: Collapse into a function here and other places?

File: tools/DocsLinter/Program.cs [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+        }
+
+        /// <summary>
+        /// A helper function to print a error when an invalid link is found.
+        /// </summary>
+        /// <param name="markdownFilePath">The path of the markdown file the error was found in.</param>
+        /// <param name="link">The link that is invalid.</param>
+        private static void LogInvalidLink(string markdownFilePath, LinkBase link)
+        {
+            Console.ForegroundColor = ConsoleColor.Red;
```
Comment (jared-improbable) [inline tools/DocsLinter/Program.cs]: Saucy.

File: tools/DocsLinter/SimplifiedMarkdownDoc.cs [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+        }
+
+        /// <summary>
+        /// A helper function that parses the Markdig link object and returns the correct type of link - remote or local. 
+        /// </summary>
+        /// <param name="linkInline">The object from Markdig to parse.</param>
+        /// <returns></returns>
+        public static LinkBase ParseLink(LinkInline linkInline)
+        {
+            if (linkInline.Url.StartsWith("http") || linkInline.Url.StartsWith("www"))
```
Comment (jared-improbable) [inline tools/DocsLinter/SimplifiedMarkdownDoc.cs]: Probably a big fragile - what about `ftp://` or sites that don't start with `www`?

File: tools/DocsLinter/ThirdPartyLicenses.md [hunk 10]
```diff
@@ -0,0 +1,25 @@
+License for [Markdig](https://github.com/lunet-io/markdig):
```
Comment (jared-improbable) [inline tools/DocsLinter/ThirdPartyLicenses.md]: Run past Legal

File: tools/DocsLinter/Tests/LinkCheckingTests.cs [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+        {
+            // Move to a temp directory
+            tempDirectory = Path.Combine(Path.GetTempPath(), "linter-docs-tests");
+            Directory.CreateDirectory(tempDirectory);
+            currentDirectory = Environment.CurrentDirectory;
+            Environment.CurrentDirectory = tempDirectory;
+
+            // Create a non-markdown file for testing purposes
+            Directory.CreateDirectory("test");
+            var fs = File.Create("test/image.png");
```
Comment (jared-improbable) [inline tools/DocsLinter/Tests/LinkCheckingTests.cs]: `File.WriteAllBytes` with an empty `byte[]`?

File: ci/test.sh (hunk 2 above)
Comment (jamiebrynes7) [reply ci/test.sh]: 🤦‍♂️

File: docs/contributions/unity-gdk-coding-standards.md (hunk 4 above)
Comment (jamiebrynes7) [reply docs/contributions/unity-gdk-coding-standards.md]: A broken link caught by the linter 😉

File: tools/DocsLinter/Program.cs (hunk 7 above)
Comment (jamiebrynes7) [reply tools/DocsLinter/Program.cs]: Yes probably a good idea

File: tools/DocsLinter/SimplifiedMarkdownDoc.cs (hunk 9 above)
Comment (jamiebrynes7) [reply tools/DocsLinter/SimplifiedMarkdownDoc.cs]: Hmmm.. I wonder if there is a generic `IsUri` function floating around

File: tools/DocsLinter/ThirdPartyLicenses.md (hunk 10 above)
Comment (jamiebrynes7) [reply tools/DocsLinter/ThirdPartyLicenses.md]: Have done, Sophie was good with putting it into a file local to the usage. Have pinged a link to the rendered markdown to get her approval.

File: tools/DocsLinter/DocsLinter.csproj (hunk 5 above)
Comment (jamiebrynes7) [reply tools/DocsLinter/DocsLinter.csproj]: Done 👍

File: tools/DocsLinter/Program.cs [hunk 12]
```diff
@@ -231,7 +223,20 @@ private static void LogLinkWarning(string markdownFilePath, LinkBase link, strin
         }
 
         /// <summary>
-        /// A helper function the collects all Markdown files from a list of file paths.
+        /// A helper function to log exceptions when they are caught.
+        /// </summary>
+        /// <param name="ex">The exception that was caught</param>
+        private static void LogException(Exception ex)
+        {
+            Console.Error.WriteLine(ex.ToString());
```
Comment (jared-improbable) [inline tools/DocsLinter/Program.cs]: Print it in red too? Or is that too much?

Comment (jamiebrynes7) [reply tools/DocsLinter/Program.cs]: Can definitely do that, I love me some console colours

File: tools/DocsLinter/ThirdPartyLicenses.md (hunk 10 above)
Comment (jamiebrynes7) [reply tools/DocsLinter/ThirdPartyLicenses.md]: Got approval, all good to go!

File: tools/DocsLinter/Program.cs [hunk 13]
```diff
@@ -0,0 +1,260 @@
...
+    /// </summary>
+    /// <param name="markdownFilePath">The fully qualified path of the Markdown file to check</param>
+    /// <param name="remoteLink">The object representing the remote link to check.</param>
+    /// <returns>A bool indicating success/failure</returns>
+    internal static bool CheckRemoteLink(string markdownFilePath, RemoteLink remoteLink)
+    {
+      try
+      {
+        var request = WebRequest.CreateHttp(remoteLink.Url);
+        request.Method = "GET";
```
Comment (samiwh) [inline tools/DocsLinter/Program.cs]: Can use HttpWebRequest.Get

File: tools/DocsLinter/Program.cs [hunk 14]
```diff
@@ -0,0 +1,260 @@
...
+        var response = request.GetResponse() as HttpWebResponse;
+        response.Close();
```
Comment (samiwh) [inline tools/DocsLinter/Program.cs]: We should close inside a finally block

File: tools/DocsLinter/SimplifiedMarkdownDoc.cs [hunk 15]
```diff
@@ -0,0 +1,204 @@
...
+      }
+
+      return new LocalLink(linkInline);
+    }
+  }
+
+  /// <summary>
+  ///   Abstract base class for a link.
+  /// </summary>
+  public abstract class LinkBase
```
Comment (samiwh) [inline tools/DocsLinter/SimplifiedMarkdownDoc.cs]: This could just be an interface

File: tools/DocsLinter/SimplifiedMarkdownDoc.cs [hunk 16]
```diff
@@ -0,0 +1,204 @@
...
+    }
+  }
+
+  /// <summary>
+  ///   A data class that represents a local link. I.e. - "../README.md#heading"
+  /// </summary>
+  public class LocalLink : LinkBase
+  {
+    /// <summary>
+    ///   The heading in a local link. May be null.
```
Comment (samiwh) [inline tools/DocsLinter/SimplifiedMarkdownDoc.cs]: Nit - these comments are probably not needed. Having a class level comment that describes what's here might be easier to read.

File: tools/DocsLinter/SimplifiedMarkdownDoc.cs [hunk 17]
```diff
@@ -0,0 +1,204 @@
...
+    public override string ToString()
+    {
+      return Url;
+    }
+  }
+
+  /// <summary>
+  ///   A data class that represents a local link. I.e. - "../README.md#heading"
+  /// </summary>
+  public class LocalLink : LinkBase
```
Comment (samiwh) [inline tools/DocsLinter/SimplifiedMarkdownDoc.cs]: If LinkBase is a interface then can this be a struct as this just looks like it's data.

File: tools/DocsLinter/SimplifiedMarkdownDoc.cs [hunk 18]
```diff
@@ -0,0 +1,204 @@
...
+    }
+  }
+
+  /// <summary>
+  ///   A  data struct that represents a Markdown heading.
+  /// </summary>
+  public struct Heading
+  {
+    /// <summary>
+    ///   Regex used in the sanitization of a Markdown title
```
Comment (samiwh) [inline tools/DocsLinter/SimplifiedMarkdownDoc.cs]: Name and the comment are the same here.
